mmdet/models/detectors/rtdetr.py

```diff
@@ -0,0 +1,354 @@
+                has shape (bs, num_feat_points). Defaults to None.
+            dn_mask (Tensor, optional): The attention mask to prevent
+                information leakage from different denoising groups and
+                matching parts, will be used as `self_attn_mask` of the
+                `self.decoder`, has shape (num_queries_total,
+                num_queries_total).
+                It is `None` when `self.training` is `False`.
+
+        Returns:
+            dict: The dictionary of decoder outputs, which includes the
+            `out_logits` and `out_bboxes` of the decoder output.
+        """
+        out_logits, out_bboxes = self.decoder(
+            query=query,
+            value=memory,
+            key_padding_mask=memory_mask,
+            self_attn_mask=dn_mask,
+            reference_points=reference_points,
+            spatial_shapes=spatial_shapes,
+            level_start_index=level_start_index,
+            valid_ratios=valid_ratios,
+            reg_branches=self.bbox_head.reg_branches,
+            cls_branches=self.bbox_head.cls_branches)
+
+        decoder_outputs_dict = dict(
+            hidden_states=out_logits, references=out_bboxes)
+        return decoder_outputs_dict
+
+    def generate_proposals(self,
+                           spatial_shapes: Tensor = None,
+                           device: Optional[torch.device] = None,
+                           grid_size: float = 0.05) -> Tuple[Tensor, Tensor]:
+        """Generate proposals from spatial shapes.
+
+        Args:
+            spatial_shapes (Tensor): Spatial shapes of features in all levels,
+                has shape (num_levels, 2), last dimension represents (h, w).
+                Defaults to None.
+            device (str | torch.device): The device where the anchors will be
+                put on. Defaults to None.
+            grid_size (float): The grid size of the anchors. Defaults to 0.05.
+
+        Returns:
+            tuple: A tuple of proposals and valid masks.
+
+            - proposals (Tensor): The proposals of the detector, has shape
+                (bs, num_proposals, 4).
+            - valid_masks (Tensor): The valid masks of the proposals, has shape
+                (bs, num_proposals).
+        """
+
+        if spatial_shapes is None:
+            spatial_shapes = [[
+                int(self.eval_size[0] / s),
+                int(self.eval_size[1] / s)
+            ] for s in self.feat_strides]
+
+        proposals = []
+        for lvl, (H, W) in enumerate(spatial_shapes):
+            grid_y, grid_x = torch.meshgrid(
+                torch.linspace(
+                    0, H - 1, H, dtype=torch.float32, device=device),
+                torch.linspace(
+                    0, W - 1, W, dtype=torch.float32, device=device))
+            grid = torch.cat([grid_x.unsqueeze(-1), grid_y.unsqueeze(-1)], -1)
+
+            valid_wh = torch.tensor([H, W], dtype=torch.float32, device=device)
```

Comment:
we need to change this line to
```python
valid_wh = torch.tensor([W, H], dtype=torch.float32, device=device)

otherwise I got problem for non square image.
This bug also is in original code, so I made issue about it
https://github.com/PaddlePaddle/PaddleDetection/issues/8680